appx.c

```diff
@@ -0,0 +1,2515 @@
+
+	uint8_t *blob = data + 4;
+
+	return d2i_PKCS7(NULL, (const unsigned char **)&blob, dataSize - 4);
+}
+
+bool appx_remove_ct_signature_entry(zipFile_t *zip, zipCentralDirectoryEntry_t *entry)
+{
+	uint8_t *data;
+	uint64_t dataSize;
+
+	if (!zipReadFileData(zip, entry, &data, &dataSize, true))
+	{
+		return false;
+	}
+
+	char *cpos = strstr((char *)data, SIGNATURE_CONTENT_TYPES_ENTRY);
+
+	if (!cpos)
+	{
+		//do not treat as en error
+		printf("Did not find existing signature entry in %s\n", entry->fileName);
+		return true;
+	}
+
+	int ipos = cpos - (char *)data;
+	int len = strlen(SIGNATURE_CONTENT_TYPES_ENTRY);
+
+	memcpy(data + ipos, data + ipos + len, dataSize - ipos - len);
+
+	dataSize -= len;
+
+	bool ret = zipOverrideFileData(zip, entry, data, dataSize, true);
+
+	free(data);
+
+	return ret;
+}
+
+bool appx_append_ct_signature_entry(zipFile_t *zip, zipCentralDirectoryEntry_t *entry)
+{
+	uint8_t *data;
+	uint64_t dataSize;
+
+	if (!zipReadFileData(zip, entry, &data, &dataSize, true))
+	{
+		return false;
+	}
+
+	char *existingEntry = strstr((char *)data, SIGNATURE_CONTENT_TYPES_ENTRY);
+
+	if (existingEntry)
+	{
+		//do not append it twice
+		return true;
+	}
+
+	char *cpos = strstr((char *)data, SIGNATURE_CONTENT_TYPES_CLOSING_TAG);
+
+	if (!cpos)
+	{
+		printf("%s parsing error\n", entry->fileName);
+		return false;
+	}
+
+	int ipos = cpos - (char *)data;
+
+	int len = strlen(SIGNATURE_CONTENT_TYPES_ENTRY);
+
+	uint64_t newSize = dataSize + len;
+	uint8_t *newData = malloc(newSize);
+
+	memcpy(newData, data, ipos);
+	memcpy(newData + ipos, SIGNATURE_CONTENT_TYPES_ENTRY, len);
+	memcpy(newData + ipos + len, data + ipos, dataSize - ipos);
+
+	bool ret = zipOverrideFileData(zip, entry, newData, newSize, true);
+
+	free(data);
+	free(newData);
+
+	return ret;
+}
+
+int appx_remove_pkcs7(FILE_FORMAT_CTX *ctx, BIO *hash, BIO *outdata)
+{
+	zipFile_t *zip = ctx->appx_ctx->zip;
+
+	zipCentralDirectoryEntry_t *entry = zipGetCDEntryByName(zip, CONTENT_TYPES_FILENAME);
+
+	if (!entry)
+	{
+		printf("Not a valid .appx file: content types file missing\n");
+		return -1;
+	}
+
+	if (!appx_remove_ct_signature_entry(zip, entry))
+	{
+		return -1;
+	}
+
+	for (zipCentralDirectoryEntry_t *entry = zip->centralDirectoryHead; entry != NULL; entry = entry->next)
+	{
+		if (strcmp(APP_SIGNATURE_FILENAME, entry->fileName))
+		{
+			uint64_t dummy;
+			if (!zipRewriteData(zip, entry, outdata, &dummy))
+			{
+				return -1;
+			}
+		}
+	}
+
+	uint64_t size = 0;
+
+	int64_t cdOffset = BIO_tell(outdata);
+
+	appx_write_central_directory(zip, outdata, true, cdOffset);
+
+	return 0;
+}
+
+/*
+ * Allocate and return SpcSipInfo object.
+ * [out] p: SpcSipInfo data
+ * [out] plen: SpcSipInfo data length
+ * [in] ctx: structure holds input and output data (unused)
+ * [returns] pointer to ASN1_OBJECT structure corresponding to SPC_SIPINFO_OBJID
+ */
+ASN1_OBJECT *appx_spc_sip_info_get(u_char **p, int *plen, FILE_FORMAT_CTX *ctx)
+{
+	const u_char appxuuid[] = {

```

Comment:
The UUID is different for bundles (B3585F0F-DEAA-9A4B-A434-95742D92ECEB), in Jsign the UUID is chosen depending on the presence of an `AppxBundleManifest.xml` entry.

Reply:
Fixed, thank you